### ISSUE TITLE
High Availability of Kube-apiserver

### ISSUE THREAD

**Issue by magicwang-cn** [kubernetes/kubernetes#19816]
i have saw the k8s's High Availability [proposals](https://github.com/kubernetes/kubernetes/blob/master/docs/proposals/high-availability.md), can we use **Active-Active**  instead of  **Warm Standby**?
if not, what's the reason?

i have a test with multi kube-apiserver, seems works fine, see as follows:

- 50 nodes
- 2 kube-apiserver instances with front proxy to talk to nodes:
![multi-apiserver](https://cloud.githubusercontent.com/assets/7587313/12413045/b9f4b480-bec6-11e5-9ee0-5c6355b1bea2.png)

**Comment by magicwang-cn** [comment]
@davidopp @lavalamp @gmarek

**Comment by AdoHe** [comment]
I think <b>Active-Active</b> is not that easy since this requires data replication happens through system capability and be bi-directional.

**Comment by magicwang-cn** [comment]
@AdoHe thank you for your reply.
could you have a specific description?

> this requires data replication happens through system capability and be bi-directional.

**Comment by gmarek** [comment]
cc @thockin

I think the question is why would you need an Active-Active. It's harder (everything gets harder if you're trying to distribute the logic - e.g. schedulers on both machines need to see the same state of cluster, and we want to somehow avoid huge number of collisions, somehow handle network partition and rejoining API servers) and I'm not sure it gives us enough to justify writing more complex code.

In other words: we can have Active-Active, but I'm sure we want to.

**Comment by davidopp** [comment]
A couple of clarifications:

1) We actually have already implemented HA. It is described here
https://github.com/kubernetes/kubernetes/blob/master/docs/admin/high-availability.md
We should remove the proposal doc you found (or at least mark it deprecated with a pointer to the doc I mentioned); sorry for the confusion.

2) For our master components, terms like active-active and hot/warm standby don't easily apply because all of the master components are essentially stateless -- that is, they store all of their state in etcd and not in any kind of local storage. When you start the Kubernetes master in HA configuration using the instructions in the doc I mentioned, you get replicated API servers (you can send reads and writes to any API server), one scheduler (with multiple standbys), and one controller manager (with multiple standbys). So the API server is already "active-active" in some sense, though like I said this term isn't quite appropriate here. We recently added support for running multiple simultaneous schedulers (#17865) so it should be possible to run multiple schedulers in parallel, but we haven't tested it yet particularly in scenarios where multiple schedulers are scheduling non-disjoint sets of pods. Replicated controller manager is tricky even though it is stateless, and we might eventually support it, but we haven't heard any compelling use case for it so this is not a high priority.

**Comment by magicwang-cn** [comment]
@davidopp Great! 
I see. What i asked is just about kube-apiserver. I know it is tricky to support active-active of controller-manager and scheduler, tks.

**Comment by magicwang-cn** [comment]
i have tested with replicated API servers, one scheduler and one controller-manager what you mentioned.
and find the problem:
nodes can not be ready completely, and from the log of kubelet, i see sometimes kubelet can not dial a tcp connection with haproxy(`just timeout`)

**Comment by mqliang** [comment]
One problem is that we may need enable ***quorum read*** when apiserver send request to etcd, otherwise ControllerManager/Scheduler/Kubelet may get old data. 

We deployment k8s as follows to make api-server high availabe.
```
  |--------|			|--------|			  |--------|              
  | etcd-1 |			| etcd-2 |		   	  | etcd-3 | 	   	
  |--------|			|--------|			  |--------|
	  |						|					  |
	  |						|					  |
	  |						|					  |
	  |						|					  |
|-------------|		 |-------------|		|-------------|	            
| apiserver-1 |		 | apiserver-2 |		| apiserver-3 | 	   	
|-------------|		 |-------------|	    |-------------|	  
	  |						|					  |
	  |						|					  |
	  |						|					  |
	  |						|					  |
|---------------------------------------------------------|	            
| 						Nginx							  | 	   	
|---------------------------------------------------------|
 	  |						|					  |
	  |						|					  |
	  |						|					  |
	  |						|					  |
|------------|		 |------------|		   |-------------|	            
|  scheduler |		 | controller |		   |    node     | 	   	
|------------|		 |------------|	       |-------------|	   		
```

But we found that ControllerManager/Scheduler/Kubelet may get old data since apiserver doesn't quorum read etcd. To solve this problem we add an option to api-server to enable quorum read, but it make read less effective.  

I am not sure such a solution is fine to community. I am glad to upstream this to community if it looks good.

**Comment by dalanlan** [comment]
+1 for linearity. I guess it wont bite anyone to enable quorum read if the number of etcd is limited.

**Comment by magicwang-cn** [comment]
@mqliang can you share your nginx configuration? and what's your k8s cluster version? thank you.

**Comment by mqliang** [comment]
@magicwang-cn I am sorry I could not share our nginx configuration here. As far the k8s cluster version, we use V1.0. But I read the latest source code of k8s, apiserver does ***NOT*** quorum read by default.

There are two cases where we may read old data if we have multi api-server:
1) When a etcd member is isolated from etcd cluster, we can still read from it(but all the write request sent to the isolated member will failed). If unfortunately nginx forward some read request to the isolated member, those read request will get old data.
2)  Consider the following case:
* A write request(e.g. ?key=hello&value=world) was forward to apiserver-1 and then forward to etcd-1, etcd-1 will return 200 indicating the write request was succeeded if the data has been written to a majority of members.(for example, written to etcd-1 and etcd-2)
* Almost immediately, a read request(?key=hello) was forward to etcd-3, if api-server doesn't enable quorum read, this read request will failed(key not found) since the data has not been synced to etcd-3.

**Comment by AdoHe** [comment]
@magicwang-cn  in Active-Active schema, we have to make sure the view of the cluster in a consistent way, for example the scheduler in different machines have to see the same state of the cluster.  For etcd since it use majority based leader election, if you don't use quorum read, you may get stale read. I think @xiang90 knows more detail.

**Comment by magicwang-cn** [comment]
@davidopp @gmarek

**Comment by gmarek** [comment]
Etcd is only one part of the equation, second one is the cache on top of etcd we're keeping in API servers and serve some requests from it. As long as there's a single write point everything should be OK, but I'm not sure how it'll work it we allow all instances to change etcd's state. It's probably OK, as it's supposed to be only an optimization. @wojtek-t @timothysc

**Comment by timothysc** [comment]
@magicwang-cn We do active-active on the api-servers today and configure our kubelets to talk through a VIP connected to a load-balancer (HA-Proxy).  

Active-passive also exists for the scheduler+controller-manager via a couple of implementations atm, but we are rallying around 1: https://github.com/kubernetes/kubernetes/issues/18553

**Comment by timothysc** [comment]
IMHO we should close this, as it's a non-issue.  

/cc @lavalamp

**Comment by lavalamp** [comment]
Yes, I'm quite confused by the OP-- multiple active apiservers is explicitly how apiserver scales. The warm standby stuff is only for controller-manager and scheduler.

**Comment by dalanlan** [comment]
This one could be safely closed, and i filed another to enable quorum read for etcd(which is an issue).

**Comment by abel-von** [comment]
@mqliang  I encountered the same problem with what you mentioned, can you post you nginx config?

**Comment by timothysc** [comment]
This is a pretty big issue honestly, at the time I didn't ruminate on it... but now that I've seen some experiments, it exposes a fundamental fault in the notion of horizontally scalable api-servers.

There are race conditions that do occur on reads, that force a quorum read.  If you enforce a quorum read to *greatly* decrease read throughput and you also force a constraint on the api-server and clients.

In days gone by, the api-servers performance used to be issue, and horizontal scalability was an escape hatch.  However, this exposed a data-consistency problem in the design.  Now-a-days the api-server is no longer a bottleneck, and in fact, we've seem pretty low cpu-profiles when it stands on it's own.  

In the coming days I'm going to try to draft up a document that rethinks how we do HA given the current state.  

/cc @wojtek-t @fgrzadkowski @lavalamp
